src/single_random_draw.rs

```diff
@@ -0,0 +1,488 @@
+///
+fn get_effective_value(
+    utxo: &TxOut,
+    input: &TxIn,
+    fee_rate: FeeRate,
+) -> Result<Option<Amount>, Error> {
+    let weight_predict = transaction::InputWeightPrediction::new(
+        input.script_sig.len(),
+        input.witness.iter().map(|elem| elem.len()),
+    );
+
+    let input_weight = transaction::predict_weight(Some(vec![weight_predict]), None::<Vec<_>>);
+    let input_fee: Option<Amount> = fee_rate.checked_mul_by_weight(input_weight);
+
+    match input_fee {
+        Some(f) => Ok(utxo.value.checked_sub(f)),
+        None => Err(Error::Multiplication(input_weight, fee_rate)),
+    }
+}
+
+/// Calculate the fee for a transaction with two UTXOs and no inputs.
+///
+/// Returns `Ok(Amount)` where Amount is the fee.
+///
+/// ### Errors
+///
+/// Returns `Err(Error::Multiplication)` if `FeeRate` * `Weight` overflows.
+///
+fn calculate_fee(
+    script_pubkey: &ScriptBuf,
+    fee_rate: FeeRate,
+    transaction: &mut Transaction,
+) -> Result<Amount, Error> {
+    let output: TxOut = TxOut { value: Amount::ZERO, script_pubkey: script_pubkey.clone() };
+
+    let output_two = output.clone();
+
+    transaction.output = vec![output, output_two];
+
+    let weight: Weight = transaction.weight();
+
+    let transaction_fee: Option<Amount> = fee_rate.checked_mul_by_weight(weight);
+
+    match transaction_fee {
+        Some(f) => Ok(f),
+        None => Err(Error::Multiplication(weight, fee_rate)),
+    }
+}
+
+/// Randomly select coins for the given target by shuffling the UTXO pool and
+/// taking UTXOs until the given target is reached.
+///
+/// The fee_rate can have an impact on the selection process since the fee
+/// must be paid for in addition to the target.  However, the total fee
+/// is dependant on the number of UTXOs consumed and the new inputs created.
+/// The selection strategy therefore calculates the fees of what is known
+/// ahead of time (the number of UTXOs create and the transaction header),
+/// and then then for each new input, the effective_value is tracked which
+/// deducts the fee for each individual input at selection time.  For more
+/// discussion see the following:
+///
+/// https://bitcoin.stackexchange.com/questions/103654/calculating-fee-based-on-fee-rate-for-bitcoin-transaction/114847#114847
+///
+/// ## Parameters
+/// ///
+/// /// * `target` - target value to send to recipient.
+/// /// * `fee_rate` - ratio of transaction amount per size.
+/// /// * `utxos` - UTXOs from which to sum the target amount.
+/// /// * `inputs` - the order corresponds to the ordering of the UTXOs where the first input consumes the first UTXO and so on.
+/// /// * `transaction` - the transaction to populate with resulting inputs and outputs.
+/// /// * `script_pubkey` - needed to compose the transaction outputs.
+/// /// * `rng` - used primarily by tests to make the selection deterministic.
+///
+pub fn select_coins_srd<R: rand::Rng + ?Sized>(
```

Comment:
The API looks more like `fund_transaction` than `select_coins`.

Reply:
I see your point.  Although I needed a way to compute the weight of the transaction that the coins might need to be included in (transaction headers).  Instead I could just expect that the weight is passed as an argument to select_coins.

Reply:
Personally I'd find more useable an interface that just takes a target value which excludes the fee for selected inputs but includes everything else. For example in rust-dlc the fee for the fund transaction is split between both parties, so this interface would be difficult to use.

Reply:
I see.  Sure we can try that and see how it looks.

Reply:
> Personally I'd find more usable an interface that just takes a target value which excludes the fee for selected inputs but includes everything else.

Besides calculating the fee, I also use the transaction weight to check that the max transaction weight hasn't been exceeded [here](https://github.com/p2pderivatives/rust-bitcoin-coin-selection/pull/23/files#diff-ce2c6b6b3132bf7b26a80fc104adfdeb0eb023b9bfd487c0da95d83830a73a32R133)

Would passing in a "weight so far" as an argument be possible?

Reply:
I'm not sure it's a useful enough feature to warrant a more complicated API personally

Reply:
I think there is a miss communication here.  You are suggesting incorporating the fee for other parts of the transaction into the target amount.  How then do we know what the weight so far is so that we can guard against the inputs exceeding the weight max when they are selected?

Reply:
I think users can check by themselves after, but if you feel like it's better to have it then go with it!

Reply:
Good point.  Instead of having each algo check that max weight hasn't been exceeded, it could just be checked afterwords.

Reply:
Ok I gutted all the transaction stuff.

Reply:
@murchandamus I'm not sure if you have an opinion but @Tibo-lg thought to just include the fee in the target instead of how core does it here: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/coinselection.cpp#L204

Reply:
It seems like there are no further comments or objections here.  Marking as resolved.